subsys/modbus/mb_rtu_client.c

```diff
@@ -0,0 +1,614 @@
+		coil_val = MODBUS_COIL_ON_CODE;
+	}
+
+	ctx->tx_frame.length = 4;
+	sys_put_be16(start_addr, &ctx->tx_frame.data[0]);
+	sys_put_be16(coil_val, &ctx->tx_frame.data[2]);
+
+	err = mbm_send_cmd(ctx, node_addr, 5, NULL);
+	k_mutex_unlock(&ctx->iface_lock);
+
+	return err;
+}
+
+int modbus_rtu_request_fc06(const int iface,
+			    const uint8_t node_addr,
+			    const uint16_t start_addr,
+			    const uint16_t reg_val)
+{
+	struct mb_rtu_context *ctx = mb_get_context(iface);
+	int err;
+
+	if (ctx == NULL) {
+		return -ENODEV;
+	}
+
+	k_mutex_lock(&ctx->iface_lock, K_FOREVER);
+
+	ctx->tx_frame.length = 4;
+	sys_put_be16(start_addr, &ctx->tx_frame.data[0]);
+	sys_put_be16(reg_val, &ctx->tx_frame.data[2]);
+
+	err = mbm_send_cmd(ctx, node_addr, 6, NULL);
+	k_mutex_unlock(&ctx->iface_lock);
+
+	return err;
+}
+
+int modbus_rtu_request_fc08(const int iface,
+			    const uint8_t node_addr,
+			    const uint16_t sfunc,
+			    const uint16_t data,
+			    uint16_t *const data_out)
+{
+	struct mb_rtu_context *ctx = mb_get_context(iface);
+	int err;
+
+	if (ctx == NULL) {
+		return -ENODEV;
+	}
+
+	k_mutex_lock(&ctx->iface_lock, K_FOREVER);
+
+	ctx->tx_frame.length = 4;
+	sys_put_be16(sfunc, &ctx->tx_frame.data[0]);
+	sys_put_be16(data, &ctx->tx_frame.data[2]);
+
+	err = mbm_send_cmd(ctx, node_addr, 8, data_out);
+	k_mutex_unlock(&ctx->iface_lock);
+
+	return err;
+}
+
+int modbus_rtu_request_fc15(const int iface,
+			    const uint8_t node_addr,
+			    const uint16_t start_addr,
+			    uint8_t *const p_coil_tbl,
+			    const uint16_t nbr_coils)
+{
+	struct mb_rtu_context *ctx = mb_get_context(iface);
+	uint8_t nbr_bytes;
+	int err;
+	uint8_t *p_data;
+
+	if (ctx == NULL) {
+		return -ENODEV;
+	}
+
+	k_mutex_lock(&ctx->iface_lock, K_FOREVER);
+
+	sys_put_be16(start_addr, &ctx->tx_frame.data[0]);
+	sys_put_be16(nbr_coils, &ctx->tx_frame.data[2]);
+
+	nbr_bytes = (uint8_t)(((nbr_coils - 1) / 8) + 1);
+	ctx->tx_frame.data[4] = nbr_bytes;
+	p_data = &ctx->tx_frame.data[5];
+	ctx->tx_frame.length = 5 + nbr_bytes;
+
+	for (uint8_t i = 0; i < nbr_bytes; i++) {
+		*p_data++ = p_coil_tbl[i];
+	}
+
+	err = mbm_send_cmd(ctx, node_addr, 15, NULL);
+	k_mutex_unlock(&ctx->iface_lock);
+
+	return err;
+}
+
+int modbus_rtu_request_fc16(const int iface,
+			    const uint8_t node_addr,
+			    const uint16_t start_addr,
+			    uint16_t *const p_reg_tbl,
+			    const uint16_t nbr_regs)
+{
+	struct mb_rtu_context *ctx = mb_get_context(iface);
+	uint8_t nbr_bytes;
+	int err;
+	uint8_t *p_data;
+
+	if (ctx == NULL) {
+		return -ENODEV;
+	}
+
+	k_mutex_lock(&ctx->iface_lock, K_FOREVER);
+
+	sys_put_be16(start_addr, &ctx->tx_frame.data[0]);
+	sys_put_be16(nbr_regs, &ctx->tx_frame.data[2]);
+
+	nbr_bytes = (uint8_t) (nbr_regs * sizeof(uint16_t));
+	ctx->tx_frame.length =  nbr_bytes + 5;
+	ctx->tx_frame.data[4] = nbr_bytes;
+	p_data = &ctx->tx_frame.data[5];
+
+	for (int i = 0; i < nbr_regs; i++) {
+		sys_put_be16(p_reg_tbl[i], p_data);
+		p_data += sizeof(uint16_t);
+	}
```

Comment:
This code block is used several times. Maybe it makes senes to create a `memcpy_be()` function.